Create control point classes

mikee47 opened the pull request:
Wrap all boilerplate stuff in classes, add `Basic_ControlPoint` sample to demonstrate. Applications don't need to touch XML at all.

To do:

* [x] Generate device/service class header from XML document tree. Documents must be organised as on target device.
* [x] Implement `ControlPoint::beginSearch` methods for pre-defined device classes. Does not require description fetch.
* [x] Implement action processing (SOAP stuff)
* [x] Add back description fetching stage so we get correct friendlyName and endpoint URLs.
* [x] Use a custom stream class to process description in chunks so we don't need so much memory.
* [x] Add `allowedValue`s as constants
* [x] Check/fix handling of SOAP errors from remote endpoint
* [x] Add `make` target for scanning local network and pulling in device description trees. Can do this very easily as a host application with recursive fetches for all description files, writing them out to directory tree. There are existing tools to do this but harder to integrate.
* [x] Organise generated directory structure and namespaces to match URN
* [x] Refactor for consistency with hosted devices.
* [x] Implement support for asynchronous action request handling
* [x] Add generator for hosted devices and services. Two files will be generated:
    - [x] The first file will contain class templates for devices and services.  Its purpose is to translate SOAP requests into typed methods.
    - [x] The second file will implement concrete clases using the templates, and must be edited by implementor to respond to requests.
* [ ] Document
* [ ] test

Note: Added python dependency for `lxml`.

The way this is going we follow these steps to implement support for any UPnP device:

1. Perform network scan to pull in device description files into hierarchical directory structure.
2. Update project component.mk file to build header file (s)
3. #include appropriate header files in application
4. Issue search on pre-defined device or service class
5. Handle found device in callback with steps such as:
   - Verifying device/service fields to see if it's the actual device we want (can cancel the search if so)
   - Storing the device object for later use
   - Issuing commands the device (calling action methods on Service object)

Having the class information pre-defined means there is no requirement to fetch or parse those large description files **provided** we are targeting a specific device. This does not cover the general case since we still need the device description to obtain the correct endpoints (and friendlyName plus a few other attributes).

Future work:

* Add event notification support
* Add action argument validation: state variables can have things such as allowedValue, minimum, maximum. This is perhaps something that hosted devices should do more than control devices, since the endpoint will validate.
* Add global exception/error handling callback(s) for framework with appropriate Error codes.
* Consider adding automatic search timeout, cancelling search or sending additional msearch requests. Should we add a search queue?
* Add support for large (streamable) result fields. For example, ContentDirectory results are in DIDL format and can be large. The declared data type is `string` but this would be better handled by a custom stream type to add the additional layer(s) of abstraction. This could be done with additional (custom) attributes in the schema or by using custom data types. The user API would be implemented in `ActionResponse` and `ActionRequest` classes.

mikee47 committed:
Add RenderingControl service name

slav-at-attachix committed:
If there is already a response stream specified -> don't replace it with a default one.

mikee47 committed:
Remove `ControlPoint::sendDescriptionRequest`

Not needed as we always call `requestDescription`

slav-at-attachix committed:
Simplify the usage of UPnP control points for the users.

mikee47 committed:
Fix comment headers

mikee47 committed:
Move UPnP initialization into separate unit and allow multiple calls to initialize()

ControlPoint automatically registers itself in constructor, and unregisters itself in destructor.

mikee47 committed:
Add Basic_ControlPoint sample: compiles, not yet functional

Use lxml with Python as xslt processor

mikee47 commented:
@slaff Hope this gives you a better idea of where I'm headed!

mikee47 committed:
Fixes for Linux

mikee47 committed:
Make getField() method const

mikee47 committed:
Add service classes as DeviceClass public member variables

mikee47 committed:
ServiceUrn bugfix

mikee47 committed:
Need large description buffer

mikee47 committed:
Implement DeviceControl::getService()

mikee47 committed:
Use UPNP_VERSION (defined in SSDP library)

mikee47 committed:
Change various methods to const

mikee47 committed:
RootDeviceList -> RootDevice::List, DeviceList -> Device::List, ServiceList -> Service::List

mikee47 commented:
@slaff Latest commits require updated SSDP library.

mikee47 committed:
Add Python requirements

mikee47 committed:
Fixed issue in SSDP with sending requests (requires second UDP connection)

mikee47 committed:
Add Usn class

mikee47 committed:
Move Device/Service classe inside Control class definitions as `Class`

slaff commented:
> Hope this gives you a better idea of where I'm headed!

You are building the dream of every creator of such semi-selfdescribing services ;)

mikee47 committed:
Declare class properies as const members, add service accessor methods.

mikee47 committed:
ControlPoint manages list of `DeviceClass` instances, and each `DeviceClass` contains set of `ServiceClass` instances.

mikee47 committed:
Add USN verification, will ignore notification if mismatch occurs

Move Usn into SSDP library
Define Object::Version = uint8_t and add virtual `Version version()` method
Rename UUID -> Uuid

mikee47 commented:
There are tools availble to build device/service descriptions (part of Intel UPnP tools) so custom devices, etc. are easily created too!

mikee47 committed:
Rename Action unit to ActionInfo (same as class)

mikee47 committed:
Put callback after parameters

mikee47 committed:
Prevent copying of ServiceControl objects

mikee47 committed:
ControlPoint owns objects, now passes references not pointers to callbacks. Add `printTo` method to action result structs.

mikee47 committed:
Revise ActionInfo for both request and response handling

mikee47 committed:
Add Sony HG1 renderer (I don't have a viera!) so we get correct endpoint URLs.

So we _are_ going to need device descriptions to support generic controls.

mikee47 committed:
Refactor searches. Don't need to fetch device description for pre-defined classes.

mikee47 committed:
Add Device::caption() method

mikee47 commented:
@slaff We can now perform actions and responses without using any XML at all. The `ControlPoint::beginSearch` method is templated, but without templated lambdas (C++20) not sure if it can be simplified. Thoughts welcome!

slaff commented:
> We can now perform actions and responses without using any XML at all.   ... Thoughts welcome!

How can I test it? I would like to create control points for the services that my TV is offering?

mikee47 commented:
Absent documention, the sticky at the top of this PR and the `Basic_ControlPoint` sample should help. Let me know if you get stuck?!

slaff commented:
Compiling Basic_ControlPoint sample with the following command:

```
make SMING_ARCH=Host ENABLE_MALLOC_COUNT=0 DEBUG_VERBOSE_LEVEL=3 ENABLE_GDB=1
```

gives errors on linking:
```
Basic_ControlPoint: Linking out/Host/debug/firmware/app
/media/slavey/416431e8-fd46-4982-896d-5025a72dd361/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Libraries/UPnP/samples/Basic_ControlPoint/out/Host/debug/lib/clib-App-d5f788d6d1c1dd58d53d4304b370719a.a(application.o): In function `Service_RenderingControl::Class::getField(UPnP::Service::Field) const':
/media/slavey/416431e8-fd46-4982-896d-5025a72dd361/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Libraries/UPnP/samples/Basic_ControlPoint/include/device/hg1.h:26: undefined reference to `Service_RenderingControl::Class::__fstr__domain'
/media/slavey/416431e8-fd46-4982-896d-5025a72dd361/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Libraries/UPnP/samples/Basic_ControlPoint/include/device/hg1.h:28: undefined reference to `Service_RenderingControl::Class::__fstr__type'
/media/slavey/416431e8-fd46-4982-896d-5025a72dd361/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Libraries/UPnP/samples/Basic_ControlPoint/include/device/hg1.h:32: undefined reference to `Service_RenderingControl::Class::__fstr__serviceId'
/media/slavey/416431e8-fd46-4982-896d-5025a72dd361/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Libraries/UPnP/samples/Basic_ControlPoint/include/device/hg1.h:34: undefined reference to `Service_RenderingControl::Class::__fstr__SCPDURL'
....
```

I am using the following versions of SSDP and UPnP:
```
+ceeaca9fc08cd3345ef06029d8a5abd599cea681 Libraries/SSDP (heads/develop)
+94553ec333a65928b6dd60d012c8fac1a1bcdde1 Libraries/UPnP (heads/feature/code-gen)
```

The generated files are attached to this comment.
[device.zip](https://github.com/mikee47/Sming-UPnP/files/5553075/device.zip)

mikee47 commented:
What's your `make list-config` ?

slaff commented:
[slavey@/media/sla...PnP/samples/Basic_ControlPoint] make list-config 

```
Basic_ControlPoint: Invoking 'list-config' for Host (debug) architecture

** Sming build configuration **

CUSTOM_TARGETS
- check-ssl
- out/Host/debug/firmware/spiff_rom.bin
LIBS
- pthread
ARDUINO_LIBRARIES
- UPnP
SUBMODULES
- /home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Arch/Host/Components/lwip/lwip
- /home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Components/FlashString
- /home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Components/http-parser
- /home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Components/libyuarel
- /home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Components/mqtt-codec
- /home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Components/rboot/rboot
- /home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Components/spiffs/spiffs
- /home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Components/ws_parser
- /home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Libraries/RapidXML
- /home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Libraries/SSDP
- /home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Libraries/UPnP
CONFIG_VARS
- COM_SPEED_SERIAL = $(COM_SPEED)
- DEBUG_PRINT_FILENAME_AND_LINE = 0
- DEBUG_VERBOSE_LEVEL = 3
- DISABLE_SPIFFS = 1
- ENABLE_CMD_EXECUTOR = 1
- ENABLE_GDB = 1
- ENABLE_LWIPDEBUG = 0
- ENABLE_SMART_CONFIG = 
- ENABLE_SSL = 0
- ENABLE_TASK_COUNT = 
- ENABLE_WPS = 
- HTTP_SERVER_EXPOSE_NAME = 1
- HTTP_SERVER_EXPOSE_VERSION = 0
- LOCALE = 
- MQTT_NO_COMPAT = 
- RBOOT_GPIO_ENABLED = 0
- RBOOT_GPIO_SKIP_ENABLED = 0
- RBOOT_ROM0_ADDR = 0x002000
- RBOOT_ROM1_ADDR = 
- RBOOT_ROM2_ADDR = 
- RBOOT_RTC_ENABLED = 0
- RBOOT_SILENT = 0
- RBOOT_SPIFFS_0 = 0x100000
- RBOOT_SPIFFS_1 = 0x300000
- SPIFF_SIZE = 524288  
- SPI_SIZE = 4M
- SSDP_DEBUG_VERBOSE = 0
- STRING_OBJECT_SIZE = 12
- TASK_QUEUE_LENGTH = 10
- UPNP_VERSION = 1.0
- USER_CFLAGS = 
- USE_US_TIMER = 1
- WIFI_PWD = 
- WIFI_SSID = 
CACHE_VARS
- APP_NAME = app
- COM_OPTS = --raw --encoding ascii
- COM_PORT = /dev/ttyUSB0
- COM_SPEED = 115200
- ENABLE_HOST_UARTID = 
- ENABLE_LEGACY_BUILD = 0
- FLASH_BIN = $(FW_BASE)/flash.bin
- FULL_COMPONENT_BUILD = 
- GDB_CMDLINE = trap '' INT; $(GDB) -x $(GDBSTUB_DIR)/gdbcmds --args $(TARGET_OUT_0) $(CLI_TARGET_OPTIONS) --pause
- HOST_FLASH_OPTIONS = --flashfile=$(FLASH_BIN) --flashsize=$(SPI_SIZE)
- HOST_NETWORK_OPTIONS = 
- HOST_PARAMETERS = 
- HOST_UART_OPTIONS = $(addprefix --uart=,$(ENABLE_HOST_UARTID))
- HOST_UART_PORTBASE = 10000
- KILL_TERM = pkill -9 -f "$(COM_PORT) $(COM_SPEED_SERIAL)" || exit 0
- PIP_ARGS = 
- SERVER_OTA_PORT = 9999
- SPIFF_BIN = spiff_rom
- SPIFF_FILES = files
- STRICT = 
- TERMINAL = $(PYTHON) -m serial.tools.miniterm $(COM_OPTS) $(COM_PORT) $(COM_SPEED_SERIAL)
- TRACE = 
RELINK_VARS
- ENABLE_MALLOC_COUNT = 0
- ENABLE_SSL = 0
- LWIP_SERVICE_INTERVAL = 2
- RBOOT_LD_TEMPLATE = $(RBOOT_DIR)/rboot.rom0.ld
- RBOOT_ROM_0 = rom0
- RBOOT_ROM_1 = rom1
- SPIFF_FILEDESC_COUNT = 7
- SSL_DEBUG = 0
DEBUG_VARS
- APP_CFLAGS =  -DCOM_SPEED_SERIAL=$(COM_SPEED_SERIAL) -DRBOOT_SPIFFS_0=$(RBOOT_SPIFFS_0) -DRBOOT_SPIFFS_1=$(RBOOT_SPIFFS_1) -DRBOOT_INTEGRATION -DDISABLE_SPIFFS=$(DISABLE_SPIFFS) -DSPIFF_SIZE=$(SPIFF_SIZE)
- APP_LIBDIR = out/Host/debug/lib
- ARCH_BASE = /home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Arch/Host
- AWK = POSIXLY_CORRECT= awk
- BUILD_BASE = $(OUT_BASE)/build
- CLANG_FORMAT = clang-format
- CMAKE = cmake
- FW_BASE = $(OUT_BASE)/firmware
- GCC_VERSION = 5.5.0
- GDBSTUB_DIR = /home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Arch/Host/Components/gdbstub
- GIT = git
- GLOBAL_CFLAGS = -DSMING_ARCH=$(SMING_ARCH) -DPROJECT_DIR=\"$(PROJECT_DIR)\" -DSMING_HOME=\"$(SMING_HOME)\" $(USER_CFLAGS) -DENABLE_CMD_EXECUTOR=$(ENABLE_CMD_EXECUTOR) -DHTTP_SERVER_EXPOSE_NAME=$(HTTP_SERVER_EXPOSE_NAME) -DHTTP_SERVER_EXPOSE_VERSION=$(HTTP_SERVER_EXPOSE_VERSION) -DDEBUG_PRINT_FILENAME_AND_LINE=$(DEBUG_PRINT_FILENAME_AND_LINE) -DCUST_FILE_BASE=$$* -DDEBUG_VERBOSE_LEVEL=$(DEBUG_VERBOSE_LEVEL) -ggdb -DENABLE_GDB=1 -DSTRING_OBJECT_SIZE=$(STRING_OBJECT_SIZE)  -DUSE_US_TIMER
- OS = 
- OUT_BASE = out/Host/debug
- PROJECT_DIR = /media/slavey/416431e8-fd46-4982-896d-5025a72dd361/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Libraries/UPnP/samples/Basic_ControlPoint
- PYTHON = python
- RBOOT_BIG_FLASH = 0
- RBOOT_TWO_ROMS = 0
- SMING_ARCH = Host
- SMING_CXX_STD = c++17
- SMING_HOME = /home/slavey/dev/esp8266.dev.box/dev/Sming/Sming
- SMING_RELEASE = 
- TOOLS_BASE = $(SMING_HOME)/$(OUT_BASE)/tools
- UNAME = Linux
- USER_LIBDIR = $(SMING_HOME)/$(OUT_BASE)/lib
```

mikee47 commented:
Try esp quick toolchain. GCC 5.5 is a PITA. I

slaff commented:
> Try esp quick toolchain.

This is the Host architecture. ESP-Quick-Toolchain is for the ESP8266 architecture.

slaff commented:
With the old ESP8266 toolchain I get:
```
/home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Libraries/UPnP/src/ActionInfo.cpp: In member function 'bool UPnP::ActionInfo::getArg(const String&, float&, float)':
/home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Libraries/UPnP/src/ActionInfo.cpp:96:27: error: 'strtof' was not declared in this scope
  value = strtof(s, nullptr);
                           ^
```

With the new one there is a deprecated method that is used:
```
C+ /home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Libraries/UPnP/src/ControlPoint.cpp
/home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Libraries/UPnP/src/ControlPoint.cpp: In lambda function:
/home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Libraries/UPnP/src/ControlPoint.cpp:246:33: error: 'String HttpConnection::getResponseString()' is deprecated [-Werror=deprecated-declarations]
  246 |    s = client.getResponseString();
      |                                 ^
In file included from /home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Core/Network/Http/HttpServerConnection.h:15,
                 from /home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Libraries/UPnP/src/include/Network/UPnP/Object.h:27,
                 from /home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Libraries/UPnP/src/include/Network/UPnP/ControlPoint.h:23,
                 from /home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Libraries/UPnP/src/ControlPoint.cpp:21:
/home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Core/Network/Http/HttpConnection.h:126:9: note: declared here
  126 |  String getResponseString() SMING_DEPRECATED

mikee47 commented:
GCC 5.5 is pretty old. To fix this requires generating an additional `.cpp` file which I suppose I should do. I'm using GCC 9.

slaff commented:
> To fix this requires generating an additional .cpp file which I suppose I should do.

Yes, please.

mikee47 committed:
Scanner/Code generator refactoring (WIP)

mikee47 commented:
This does raise (again) the issue of how long we intend to support these ancient toolchains.

mikee47 committed:
Move Urn into SSDP library, handles both URN and USN strings

mikee47 committed:
Move ControlPoint::Search into separate header and break into virtual classes.

mikee47 committed:
Simplify, serialise http fetches

mikee47 committed:
Pass description pointer to callback, null on error

mikee47 committed:
Scan going OK.

mikee47 committed:
Nested queuing

mikee47 committed:
Add COMPONENT_PREREQUISITES and COMPONENT_BUILD_BASE to build framework

mikee47 committed:
Define proper build targets for source code generation

slaff commented:
> This does raise (again) the issue of how long we intend to support these ancient toolchains.

I guess we should start communicating this to the Sming developers/users and stop using the old Esp8266 compiler middle of 2021. 

Still I would prefer to have as a minimum requirement a compiler that supports C++11/c11 standards. There are other architectures out there and some of them have gnu-based ancient compilers.

With that said it would be great to have your changes so that I can compile the sample and test it with my TV :)

mikee47 commented:
@slaff I've started work on scanning here `https://github.com/mikee47/Sming-UPnP/tree/feature/scanning` and sort of works but really giving the TCP stack a bit of a thrashing. I'm serialising requests but getting "Out of TCP connections" error and it's stalling. Could you give it a test and see how it runs in Linux? (My VM can't see any PnP devices...)

mikee47 commented:
>> With that said it would be great to have your changes so that I can compile the sample and test it with my TV :)

If you don't want to wait just install a more recent version of GCC!

slaff commented:
> If you don't want to wait just install a more recent version of GCC!

Already done. I am using GCC 9 now.

mikee47 committed:
Output device schema by manufacturer/friendlyName/deviceType

Device implementations vary so only services are common

mikee47 commented:
@slaff The scanner now works rather better. It generates a device tree for the local network (under `devices`), and also schema information.

Because device implementations vary, they're not re-usable so output as `schema/device/{manufacturer}/{friendlyName}/{deviceType+version}.xml`. Service schema are re-useable so are namespaced into `schema/service/{domain}/{serviceType+version}.xml`.

Could you give this arrangement some thought? I'll revise the header generation to suit so the directory hierarchy for services will be similar. Devices will need manual rearangement I think, but this does most of the work.

Note that I need to accommodate embedded devices so will add a `getDevice()` method, just like `getService()`.

mikee47 commented:
Try a web search "xslt UPnP library"...

slaff commented:
> Try a web search "xslt UPnP library"...

What do you mean? To see other implementations or?

slaff commented:
> Could you give this arrangement some thought?

The arrangement looks good to me. My only suggestion would be to put these output directories under out/ so that they are ignored by git and also cleaned with `make clean`.

slaff commented:
> Could you give it a test and see how it runs in Linux? (My VM can't see any PnP devices...)

Runs ok on my Linux. I am testing it using the following:
-  created a tap0 interface without adding forwarding and nat-ing.
```bash

- 
# Create a tap0 interface with IP network 192.168.13.1/24
sudo ip tuntap add dev tap0 mode tap user $(whoami)
sudo ip a a dev tap0 192.168.13.1/24
sudo ifconfig tap0 up

# The following lines are needed if you plan to access Internet
sudo sysctl net.ipv4.ip_forward=1
sudo sysctl net.ipv6.conf.default.forwarding=1
sudo sysctl net.ipv6.conf.all.forwarding=1
```
-  After this started a test server using https://github.com/fraunhoferfokus/peer-dial. The test server runs on all network interfaces (including tap0)
- compiled and started the scanner and all seemed good. The generated directories and files were all there for me to see and use.

> (My VM can't see any PnP devices...)

This will need multicast forwarding which I was not able to set up. Should sit with a network guru and refine the iptables rules(probably remove RELATED as most of the forwarded multicast udp packages will not be RELATED).

mikee47 commented:
> 
> 
> > Try a web search "xslt UPnP library"...
> 
> What do you mean? To see other implementations or?

Third hit is for Sming :-)

mikee47 committed:
Rearrange schema by domain

mikee47 committed:
Rearrange schema structure under domain

mikee47 commented:
@slaff I've merged common service schema into the UPnP library. These are all built into the library, although as intermediate source files are created it needs a few passes for now.

Headers and sources is written to `$(SMING_HOME)/out/common` as it's identical for all architectures so only needs doing once. Build system needs tweaking to deal with this a bit more smoothly. Location for discussion, of course.

The schema processing has been overhauled quite a bit:

* There are now four (simpler) stylesheets, two each for device/service. An additional `common.xsl` deals with name parsing,etc.
* Instead of cluttering the source with aliases, all the action variable types are translated to their C++ equivalents.
* All leading `X_` and `A_ARG_TYPE_` have been removed.
* Variable names have a 'v' prefix. One reason is to ensure it's camelCase, the other is because names such as `String` cause problems.
* Everything's namespaced by domain so should be logical.

The Basic_ControlPoint application has it's own `schema` directory since I want to demonstrate how applications can add their own device schema.

Things will probably make more sense once I've added description parsing back in, I'll aim to get that done tomorrow.

mikee47 committed:
Simplify/improve scanner queues, print final summary

mikee47 committed:
Don't attempt to re-fetch descriptions if file already present

mikee47 committed:
Fix skipping of existing files, add `fetch` command to fetch description directly.

mikee47 committed:
Fetch description and use to configure device

mikee47 committed:
Tidy ContentDirectory browse result

mikee47 committed:
Use apply-templates in preference to call-template

mikee47 committed:
Pre-defined values (statevariable allowed values)

slaff commented on tools/xsl/common.xsl:
```diff
@@ -0,0 +1,164 @@
+<xsl:template name="action-result-callback"><xsl:call-template name="action-result"/>Callback</xsl:template>
+
+<!-- Method declaration for service action -->
+<xsl:template name="action-method">action_<xsl:call-template name="action-name"/>(
```
Maybe the names can be `actionSomeThing` instead of the mixed `action_SomeThing`?

slaff replied:
What were your concerns when adding the `action_` prefix. Why not just use lower-case-first-letter(methodName). Instead of `action_SomeThing` to have `someThing`?

mikee47 replied:
It would be nice to camelCase both action methods and variable names, but it does run the risk of clashing with an existing method. Same goes for types. The only reason for `action_*` vs `action*` is to highlight that these are UPnP action methods and not regular methods.

I'll push a commit with lower camel-case naming and see how things go!

(If we get any collisions can just rename the internal/private method which _hopefully_ won't be one used by applications.)

slaff replied:
> I'll push a commit with lower camel-case naming and see how things go!

Great, thanks a lot!

slaff commented on src/include/Network/UPnP/ServiceClass.h:
```diff
@@ -0,0 +1,70 @@
+	}
+
+protected:
+	virtual ServiceControl* createObject(DeviceControl& device, const ServiceClass& serviceClass) const = 0;
```
I guess you will need here also an empty virtual destructor. See https://gcc.gnu.org/faq.html#vtables.

mikee47 replied:
Are you getting compiler errors? ServiceClass inherits (eventually) from `Item` which has the virtual destructor, so should be sufficient?

slaff replied:
> Are you getting compiler errors? ServiceClass inherits

Yes. Multiple ```undefined reference to `vtable for` errors```. Compiler GCC 9. Compiling arguments:
`make SMING_ARCH=Host ENABLE_MALLOC_COUNT=0 ENABLE_GDB=1 DEBUG_VERBOSE_LEVEL=3`

GCC version:
`gcc version 9.3.0 (Ubuntu 9.3.0-10ubuntu2~16.04) `


```
Basic_ControlPoint: Linking out/Host/debug/firmware/app
/media/slavey/416431e8-fd46-4982-896d-5025a72dd361/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Libraries/UPnP/samples/Basic_ControlPoint/out/Host/debug/lib/clib-App-d5f788d6d1c1dd58d53d4304b370719a.a(application.o): In function `_ZN4UPnP16schemas_upnp_org7service17RenderingControl15ClassCI2NS_12ServiceClassEERKNS_11DeviceClassE':
/home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/out/common/Network/UPnP/schemas-upnp-org/service/RenderingControl1.h:30: undefined reference to `vtable for UPnP::schemas_upnp_org::service::RenderingControl1::Class'
/media/slavey/416431e8-fd46-4982-896d-5025a72dd361/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Libraries/UPnP/samples/Basic_ControlPoint/out/Host/debug/lib/clib-App-d5f788d6d1c1dd58d53d4304b370719a.a(application.o): In function `_ZN4UPnP16schemas_upnp_org7service18ConnectionManager15ClassCI2NS_12ServiceClassEERKNS_11DeviceClassE':
/home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/out/common/Network/UPnP/schemas-upnp-org/service/ConnectionManager1.h:30: undefined reference to `vtable for UPnP::schemas_upnp_org::service::ConnectionManager1::Class'
/media/slavey/416431e8-fd46-4982-896d-5025a72dd361/slavey/dev/esp8266.dev.box/dev/Sming/Sming/Libraries/UPnP/samples/Basic_ControlPoint/out/Host/debug/lib/clib-App-d5f788d6d1c1dd58d53d4304b370719a.a(application.o): In function `_ZN4UPnP16schemas_upnp_org7service12AVTransport15ClassCI2NS_12ServiceClassEERKNS_11DeviceClassE':
/home/slavey/dev/esp8266.dev.box/dev/Sming/Sming/out/common/Network/UPnP/schemas-upnp-org/service/AVTransport1.h:30: undefined reference to `vtable for UPnP::schemas_upnp_org::service::AVTransport1::Class'
```

mikee47 replied:
It's probably the intermediate source. Try:

```
make UPnP-build
make
```

mikee47 replied:
Need to fix that compile issue, and consider how to handle generated source within the framework.

slaff reviewed (commented):
So far looks great !

mikee47 committed:
Move `Urn` into global namespace (like Uuid).

mikee47 committed:
camelCase variable names

mikee47 commented:
I wanted to include the pre-defined state variable values. It's a little confusing because they're called `allowedValue` but it's not exclusive. If it was, I could define them as enumerations instead of strings, but we need to allow use of other, arbitrary values as well.

These go inside the related class (here, `AVTransport1`):

```
	struct TransportState {
		DEFINE_FSTR_LOCAL(fs_NO_MEDIA_PRESENT, "NO_MEDIA_PRESENT")
		DEFINE_FSTR_LOCAL(fs_STOPPED, "STOPPED")
		DEFINE_FSTR_LOCAL(fs_TRANSITIONING, "TRANSITIONING")
		DEFINE_FSTR_LOCAL(fs_PLAYING, "PLAYING")
		DEFINE_FSTR_LOCAL(fs_PAUSED_PLAYBACK, "PAUSED_PLAYBACK")
	};
```

This is why the `fs_` prefix:

```
	struct TransportPlaySpeed {
		DEFINE_FSTR_LOCAL(fs_1, "1")
		DEFINE_FSTR_LOCAL(fs_1_2, "1/2")
		DEFINE_FSTR_LOCAL(fs_2, "2")
		DEFINE_FSTR_LOCAL(fs_5, "5")
		DEFINE_FSTR_LOCAL(fs_30, "30")
		DEFINE_FSTR_LOCAL(fs_70, "70")
		DEFINE_FSTR_LOCAL(fs_100, "100")
		DEFINE_FSTR_LOCAL(fs__1_2, "-1/2")
		DEFINE_FSTR_LOCAL(fs__2, "-2")
		DEFINE_FSTR_LOCAL(fs__5, "-5")
		DEFINE_FSTR_LOCAL(fs__30, "-30")
		DEFINE_FSTR_LOCAL(fs__70, "-70")
		DEFINE_FSTR_LOCAL(fs__100, "-100")
	};

```

If you have any suggestions for avoiding this, love to hear it!

mikee47 commented:
> 
> > This does raise (again) the issue of how long we intend to support these ancient toolchains.
> 
> I guess we should start communicating this to the Sming developers/users and stop using the old Esp8266 compiler middle of 2021.
> 
> Still I would prefer to have as a minimum requirement a compiler that supports C++11/c11 standards. There are other architectures out there and some of them have gnu-based ancient compilers.

As far as this library (UPnP) is concerned, we should basically ignore older compilers as it's just making more work for ourselves and preventing use of compiler features which are there to improve code quality, easy of use, etc. That would also apply to any libraries or samples which make use of it so, for now, we just make sure they are out of the main Sming repo (which they are!).

mikee47 committed:
camelCase action names

mikee47 committed:
Fixes

mikee47 committed:
Update .gitignore

mikee47 committed:
Move schema into separate library

mikee47 committed:
Add scan tool definitions

mikee47 committed:
Remove redundant config from Basic_ControlPoint

mikee47 commented:
@slaff Build should now be working as expected. I've opened a PR in the main Sming repo. to support this https://github.com/SmingHub/Sming/pull/2154

mikee47 committed:
Add tool make targets

mikee47 committed:
Use XML::getValue()

mikee47 committed:
Configure service endpoints during description parsing

mikee47 committed:
Fix scanner - need to compare full file path not just relative path

mikee47 committed:
Revise class structures for consistency with hosted devices

Device already has devices_ and services_ so change to owned list and use it.
getService() returns pointer. ControlPoint manages classes.
Use PROGMEM structures for class registration, grouped by domain

Why do we get HPE_CLOSE_CONNECTION if callback is queued?

mikee47 committed:
Process incoming description stream in sections using DescriptionParser

mikee47 committed:
Check Search callback provided before submission & fix potential memory leak if it fails

mikee47 committed:
remove commented-out code, Fix comment headers

mikee47 committed:
Refactor and simplify to support hosted devices with schema

mikee47 committed:
Refactor class information to fix constexpr initialisation issues

mikee47 committed:
Don't emit serviceList or deviceList if empty (compliance failure)

mikee47 committed:
Sort out URL handling - everything should base off URL used for device descrption (WIP)

mikee47 committed:
Send bad request to confirm correct fault response handling

mikee47 committed:
Action processing wrappers; Change action result methods to getXX & setXX

mikee47 committed:
Add `ErrorCode` return value for `handleAction`

mikee47 committed:
Move schema generation entirely into UPnP-Schema lilbrary

All Component /schema directories picked up automatically

mikee47 committed:
Refactoring, many updates

	Add command line parameters to scanner
Add root UPnP `schema` directory for services
Fix scan so it works on both Linux and Windows
Generate separate source/header files
Write generated files to `out/`
Add `DeviceClass::operator==()`
Scanner needs to insert service details into scpd
Add `DEVICE_DIR` and `SCHEMA_DIR` environment variables for scanner, and append `serviceType` node to schema
Copy namespace attributes into written device schema
Fix missing service descriptions (SSDP message refer to device, not service)
Update README

mikee47 committed:
Add additional Device & Service class information

mikee47 committed:
Clarify Basic_ControlPoint result printing

mikee47 committed:
Simplify Wemo class

mikee47 committed:
Redefine Error codes with negative values, add Pending

mikee47 committed:
Add `ErrorCode getErrorCode(Error)`

mikee47 committed:
Action handlers return `Error`

mikee47 committed:
Rename `result` -> `response`

mikee47 committed:
Implement deferred response handling

mikee47 commented:
@slaff OK, this is pretty much feature complete for now. If there's anything in particular you'd like me to address before merging?

mikee47 committed:
Change scanner schema output to match that of UPnP-Schema

mikee47 committed:
Fix problem with // in URL

mikee47 committed:
Allow requests with no callback

mikee47 committed:
Add GUPnP network light demo to Basic_ControlPoint sample

slaff commented on samples/Basic_ControlPoint/app/application.cpp:
```diff
@@ -0,0 +1,192 @@
+
+void initUPnP()
+{
+	UPnP::schemas_upnp_org::registerClasses();
```
Maybe you can add here comments explaining why is the registerClassess needed and what it is actually doing.

slaff commented on samples/Basic_UPnP/README.rst:
```diff
@@ -11,10 +11,6 @@ Services
-:cpp:class:`Wemo::BasicEventService`.
 
+The device and service descriptions are stored in the ``schema`` directory.
+This is used by UPnP to generate class code, so all we need to do is implement the action methods.
```
I am really interested to see how this can be done. I am planning to create a sample [Sming UPnP application for an existing Sonoff TH16 switch](https://github.com/slaff/Sming/issues/64) at home.

mikee47 replied:
https://www.meshcommander.com/upnptools contains a `Service Author` program which can help with creating custom schema. It's Windows-only...

slaff reviewed (commented):
Looks good to me. Some more helping comments and  sentence(s) about the Device generation would be great. See my comments.

mikee47 committed:
Put Wemo seviceId in the schema

mikee47 committed:
Esp32 fixes

mikee47 committed:
Update docs

mikee47 committed:
Remove unused code

mikee47 committed:
Don't wipe request when converting to Response.

Here's generated dispatch code:

return static_cast<S*>(this)->setBinaryState(
	req.getArg<bool>(Arg::BinaryState),
	req); ///< This creates ActionResponse object from ActionRequest

C++ does **not** guarantee that method parameters are processed left -> right.
This works on actual hardware, but when run in Linux Host mode it fails because
the input parameter is read **after** the construction of ActionResponse.

We now just set a flag inside ActionResponse and don't wipe the request until actually necessary.
The alternative is more verbose wrapper code.
